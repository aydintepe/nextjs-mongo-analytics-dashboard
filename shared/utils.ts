export function pluralize(count: number, noun: string, suffix = "s") {
  return `${noun}${count !== 1 ? suffix : ""}`;
}



function getEntriesSortedByField<T>(
  map: Record<string | number | symbol, T>,
  field: keyof T
) {
  return Object.entries(map).sort(([keyA, valueA], [keyB, valueB]) => {
    if (
      typeof valueA[field] === "number" &&
      typeof valueB[field] === "number"
    ) {
      return (valueA[field] as number) - (valueB[field] as number);
    }

    if (
      typeof valueA[field] === "string" &&
      typeof valueB[field] === "string"
    ) {
      return (valueA[field] as string).localeCompare(valueB[field] as string);
    }

    return 0;
  });
}

/**
 * Returns object keys sorted by value field.
 * Value field shoud be number or string.
 */
export function getKeysSortedByField<T>(map: Record<string, T>, field: keyof T) {
  return getEntriesSortedByField(map, field).map(([key]) => key);
}

/**
 * Returns object keys sorted by value field.
 * Value field shoud be number or string.
 */
export function getValuesSortedByField<T>(
  map: Record<string, T>,
  field: keyof T
) {
  return getEntriesSortedByField(map, field).map(([_, value]) => value);
}
