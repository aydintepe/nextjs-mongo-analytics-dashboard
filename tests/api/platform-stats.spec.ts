import { getApiUrl } from "../utils/url";

const axios = require("axios");

describe("/api/platform-stats", () => {
  describe("/total", () => {
    it("responds with total stats for the timerange", async () => {
      const res = await axios.get(
        `${getApiUrl()}/platform-stats/total?startDate=2023-01-01T00:00:00.000Z&endDate=2023-03-10T00:00:00.000Z`
      );

      expect(res.data).toMatchSnapshot();
    });
  });

  describe("/day", () => {
    it("responds with daily stats for the timerange", async () => {
      const res = await axios.get(
        `${getApiUrl()}/platform-stats/day?startDate=2023-01-01T00:00:00.000Z&endDate=2023-03-10T00:00:00.000Z`
      );

      expect(res.data).toMatchSnapshot();
    });
  });
});
