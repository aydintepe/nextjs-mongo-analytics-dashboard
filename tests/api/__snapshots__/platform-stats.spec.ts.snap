// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`/api/platform-stats /day responds with daily stats for the timerange 1`] = `
{
  "dayStats": [
    {
      "averageOrderRevenue": 471.5,
      "date": "2023-01-01T00:00:00.000Z",
      "orders": 4,
      "platform": "Amazon",
      "revenue": 1886,
    },
    {
      "averageOrderRevenue": 275.33333,
      "date": "2023-01-02T00:00:00.000Z",
      "orders": 6,
      "platform": "Amazon",
      "revenue": 1652,
    },
    {
      "averageOrderRevenue": 321.6,
      "date": "2023-01-03T00:00:00.000Z",
      "orders": 5,
      "platform": "Allegro",
      "revenue": 1608,
    },
    {
      "averageOrderRevenue": 351,
      "date": "2023-01-03T00:00:00.000Z",
      "orders": 5,
      "platform": "Ebay",
      "revenue": 1755,
    },
    {
      "averageOrderRevenue": 171.8,
      "date": "2023-01-04T00:00:00.000Z",
      "orders": 5,
      "platform": "Ebay",
      "revenue": 859,
    },
    {
      "averageOrderRevenue": 297.66667,
      "date": "2023-01-05T00:00:00.000Z",
      "orders": 3,
      "platform": "Allegro",
      "revenue": 893,
    },
    {
      "averageOrderRevenue": 214.33333,
      "date": "2023-01-06T00:00:00.000Z",
      "orders": 6,
      "platform": "Allegro",
      "revenue": 1286,
    },
    {
      "averageOrderRevenue": 130.25,
      "date": "2023-01-06T00:00:00.000Z",
      "orders": 4,
      "platform": "Amazon",
      "revenue": 521,
    },
    {
      "averageOrderRevenue": 36,
      "date": "2023-01-06T00:00:00.000Z",
      "orders": 1,
      "platform": "Farfetch",
      "revenue": 36,
    },
    {
      "averageOrderRevenue": 296,
      "date": "2023-01-07T00:00:00.000Z",
      "orders": 4,
      "platform": "Ebay",
      "revenue": 1184,
    },
    {
      "averageOrderRevenue": 615,
      "date": "2023-01-08T00:00:00.000Z",
      "orders": 2,
      "platform": "Allegro",
      "revenue": 1230,
    },
    {
      "averageOrderRevenue": 402.75,
      "date": "2023-01-10T00:00:00.000Z",
      "orders": 8,
      "platform": "Allegro",
      "revenue": 3222,
    },
    {
      "averageOrderRevenue": 345.4,
      "date": "2023-01-10T00:00:00.000Z",
      "orders": 5,
      "platform": "Ebay",
      "revenue": 1727,
    },
    {
      "averageOrderRevenue": 239.11111,
      "date": "2023-01-11T00:00:00.000Z",
      "orders": 9,
      "platform": "Amazon",
      "revenue": 2152,
    },
    {
      "averageOrderRevenue": 255.66667,
      "date": "2023-01-11T00:00:00.000Z",
      "orders": 3,
      "platform": "Ebay",
      "revenue": 767,
    },
    {
      "averageOrderRevenue": 278.66667,
      "date": "2023-01-12T00:00:00.000Z",
      "orders": 9,
      "platform": "Ebay",
      "revenue": 2508,
    },
    {
      "averageOrderRevenue": 64,
      "date": "2023-01-12T00:00:00.000Z",
      "orders": 1,
      "platform": "Farfetch",
      "revenue": 64,
    },
    {
      "averageOrderRevenue": 260.57143,
      "date": "2023-01-15T00:00:00.000Z",
      "orders": 7,
      "platform": "Ebay",
      "revenue": 1824,
    },
    {
      "averageOrderRevenue": 39.25,
      "date": "2023-01-16T00:00:00.000Z",
      "orders": 4,
      "platform": "Allegro",
      "revenue": 157,
    },
    {
      "averageOrderRevenue": 352.7,
      "date": "2023-01-16T00:00:00.000Z",
      "orders": 10,
      "platform": "Farfetch",
      "revenue": 3527,
    },
    {
      "averageOrderRevenue": 368,
      "date": "2023-01-17T00:00:00.000Z",
      "orders": 4,
      "platform": "Allegro",
      "revenue": 1472,
    },
    {
      "averageOrderRevenue": 127.25,
      "date": "2023-01-19T00:00:00.000Z",
      "orders": 8,
      "platform": "Farfetch",
      "revenue": 1018,
    },
    {
      "averageOrderRevenue": 223.4,
      "date": "2023-01-21T00:00:00.000Z",
      "orders": 10,
      "platform": "Amazon",
      "revenue": 2234,
    },
    {
      "averageOrderRevenue": 273.26316,
      "date": "2023-01-21T00:00:00.000Z",
      "orders": 19,
      "platform": "Ebay",
      "revenue": 5192,
    },
    {
      "averageOrderRevenue": 168,
      "date": "2023-01-22T00:00:00.000Z",
      "orders": 9,
      "platform": "Amazon",
      "revenue": 1512,
    },
    {
      "averageOrderRevenue": 157,
      "date": "2023-01-22T00:00:00.000Z",
      "orders": 4,
      "platform": "Ebay",
      "revenue": 628,
    },
    {
      "averageOrderRevenue": 193.16667,
      "date": "2023-01-23T00:00:00.000Z",
      "orders": 6,
      "platform": "Allegro",
      "revenue": 1159,
    },
    {
      "averageOrderRevenue": 105.16667,
      "date": "2023-01-24T00:00:00.000Z",
      "orders": 6,
      "platform": "Allegro",
      "revenue": 631,
    },
    {
      "averageOrderRevenue": 292.2,
      "date": "2023-01-25T00:00:00.000Z",
      "orders": 10,
      "platform": "Ebay",
      "revenue": 2922,
    },
    {
      "averageOrderRevenue": 203,
      "date": "2023-01-26T00:00:00.000Z",
      "orders": 8,
      "platform": "Farfetch",
      "revenue": 1624,
    },
    {
      "averageOrderRevenue": 285.33333,
      "date": "2023-01-27T00:00:00.000Z",
      "orders": 3,
      "platform": "Allegro",
      "revenue": 856,
    },
    {
      "averageOrderRevenue": 380.66667,
      "date": "2023-01-28T00:00:00.000Z",
      "orders": 3,
      "platform": "Allegro",
      "revenue": 1142,
    },
    {
      "averageOrderRevenue": 302.66667,
      "date": "2023-01-28T00:00:00.000Z",
      "orders": 9,
      "platform": "Ebay",
      "revenue": 2724,
    },
    {
      "averageOrderRevenue": 338.4,
      "date": "2023-01-29T00:00:00.000Z",
      "orders": 5,
      "platform": "Allegro",
      "revenue": 1692,
    },
    {
      "averageOrderRevenue": 215.5,
      "date": "2023-01-29T00:00:00.000Z",
      "orders": 10,
      "platform": "Amazon",
      "revenue": 2155,
    },
    {
      "averageOrderRevenue": 213.33333,
      "date": "2023-01-31T00:00:00.000Z",
      "orders": 3,
      "platform": "Ebay",
      "revenue": 640,
    },
    {
      "averageOrderRevenue": 460.3,
      "date": "2023-02-02T00:00:00.000Z",
      "orders": 10,
      "platform": "Ebay",
      "revenue": 4603,
    },
    {
      "averageOrderRevenue": 332.66667,
      "date": "2023-02-03T00:00:00.000Z",
      "orders": 3,
      "platform": "Farfetch",
      "revenue": 998,
    },
    {
      "averageOrderRevenue": 299.55556,
      "date": "2023-02-05T00:00:00.000Z",
      "orders": 9,
      "platform": "Allegro",
      "revenue": 2696,
    },
    {
      "averageOrderRevenue": 364.33333,
      "date": "2023-02-06T00:00:00.000Z",
      "orders": 3,
      "platform": "Farfetch",
      "revenue": 1093,
    },
    {
      "averageOrderRevenue": 280.375,
      "date": "2023-02-07T00:00:00.000Z",
      "orders": 8,
      "platform": "Farfetch",
      "revenue": 2243,
    },
    {
      "averageOrderRevenue": 255.83333,
      "date": "2023-02-08T00:00:00.000Z",
      "orders": 6,
      "platform": "Allegro",
      "revenue": 1535,
    },
    {
      "averageOrderRevenue": 252,
      "date": "2023-02-08T00:00:00.000Z",
      "orders": 9,
      "platform": "Amazon",
      "revenue": 2268,
    },
    {
      "averageOrderRevenue": 318.8,
      "date": "2023-02-09T00:00:00.000Z",
      "orders": 5,
      "platform": "Amazon",
      "revenue": 1594,
    },
    {
      "averageOrderRevenue": 299.27273,
      "date": "2023-02-10T00:00:00.000Z",
      "orders": 11,
      "platform": "Allegro",
      "revenue": 3292,
    },
    {
      "averageOrderRevenue": 252.8,
      "date": "2023-02-10T00:00:00.000Z",
      "orders": 10,
      "platform": "Farfetch",
      "revenue": 2528,
    },
    {
      "averageOrderRevenue": 218.70588,
      "date": "2023-02-11T00:00:00.000Z",
      "orders": 17,
      "platform": "Farfetch",
      "revenue": 3718,
    },
    {
      "averageOrderRevenue": 228.57143,
      "date": "2023-02-13T00:00:00.000Z",
      "orders": 7,
      "platform": "Allegro",
      "revenue": 1600,
    },
    {
      "averageOrderRevenue": 261.58824,
      "date": "2023-02-13T00:00:00.000Z",
      "orders": 17,
      "platform": "Farfetch",
      "revenue": 4447,
    },
    {
      "averageOrderRevenue": 206.88889,
      "date": "2023-02-14T00:00:00.000Z",
      "orders": 9,
      "platform": "Allegro",
      "revenue": 1862,
    },
    {
      "averageOrderRevenue": 266.4,
      "date": "2023-02-16T00:00:00.000Z",
      "orders": 5,
      "platform": "Amazon",
      "revenue": 1332,
    },
    {
      "averageOrderRevenue": 390,
      "date": "2023-02-16T00:00:00.000Z",
      "orders": 9,
      "platform": "Ebay",
      "revenue": 3510,
    },
    {
      "averageOrderRevenue": 157.5,
      "date": "2023-02-17T00:00:00.000Z",
      "orders": 10,
      "platform": "Allegro",
      "revenue": 1575,
    },
    {
      "averageOrderRevenue": 393.66667,
      "date": "2023-02-19T00:00:00.000Z",
      "orders": 6,
      "platform": "Allegro",
      "revenue": 2362,
    },
    {
      "averageOrderRevenue": 306.5625,
      "date": "2023-02-20T00:00:00.000Z",
      "orders": 16,
      "platform": "Farfetch",
      "revenue": 4905,
    },
    {
      "averageOrderRevenue": 230.77778,
      "date": "2023-02-21T00:00:00.000Z",
      "orders": 9,
      "platform": "Allegro",
      "revenue": 2077,
    },
    {
      "averageOrderRevenue": 318.8,
      "date": "2023-02-23T00:00:00.000Z",
      "orders": 10,
      "platform": "Farfetch",
      "revenue": 3188,
    },
    {
      "averageOrderRevenue": 579,
      "date": "2023-02-24T00:00:00.000Z",
      "orders": 4,
      "platform": "Allegro",
      "revenue": 2316,
    },
    {
      "averageOrderRevenue": 256.7,
      "date": "2023-02-25T00:00:00.000Z",
      "orders": 10,
      "platform": "Allegro",
      "revenue": 2567,
    },
    {
      "averageOrderRevenue": 244.1875,
      "date": "2023-02-25T00:00:00.000Z",
      "orders": 16,
      "platform": "Amazon",
      "revenue": 3907,
    },
    {
      "averageOrderRevenue": 265.25,
      "date": "2023-02-25T00:00:00.000Z",
      "orders": 16,
      "platform": "Ebay",
      "revenue": 4244,
    },
    {
      "averageOrderRevenue": 302.28571,
      "date": "2023-02-25T00:00:00.000Z",
      "orders": 7,
      "platform": "Farfetch",
      "revenue": 2116,
    },
    {
      "averageOrderRevenue": 296.92,
      "date": "2023-02-26T00:00:00.000Z",
      "orders": 25,
      "platform": "Ebay",
      "revenue": 7423,
    },
    {
      "averageOrderRevenue": 371.4,
      "date": "2023-02-27T00:00:00.000Z",
      "orders": 5,
      "platform": "Allegro",
      "revenue": 1857,
    },
    {
      "averageOrderRevenue": 354.66667,
      "date": "2023-02-27T00:00:00.000Z",
      "orders": 3,
      "platform": "Amazon",
      "revenue": 1064,
    },
    {
      "averageOrderRevenue": 257.77778,
      "date": "2023-02-27T00:00:00.000Z",
      "orders": 9,
      "platform": "Ebay",
      "revenue": 2320,
    },
    {
      "averageOrderRevenue": 68.6,
      "date": "2023-02-28T00:00:00.000Z",
      "orders": 5,
      "platform": "Ebay",
      "revenue": 343,
    },
    {
      "averageOrderRevenue": 235.25,
      "date": "2023-02-28T00:00:00.000Z",
      "orders": 4,
      "platform": "Farfetch",
      "revenue": 941,
    },
    {
      "averageOrderRevenue": 247.5,
      "date": "2023-03-01T00:00:00.000Z",
      "orders": 6,
      "platform": "Amazon",
      "revenue": 1485,
    },
    {
      "averageOrderRevenue": 138.4,
      "date": "2023-03-01T00:00:00.000Z",
      "orders": 10,
      "platform": "Ebay",
      "revenue": 1384,
    },
    {
      "averageOrderRevenue": 227.625,
      "date": "2023-03-02T00:00:00.000Z",
      "orders": 8,
      "platform": "Farfetch",
      "revenue": 1821,
    },
    {
      "averageOrderRevenue": 300.5,
      "date": "2023-03-03T00:00:00.000Z",
      "orders": 4,
      "platform": "Allegro",
      "revenue": 1202,
    },
    {
      "averageOrderRevenue": 170.4,
      "date": "2023-03-03T00:00:00.000Z",
      "orders": 5,
      "platform": "Amazon",
      "revenue": 852,
    },
    {
      "averageOrderRevenue": 314.33333,
      "date": "2023-03-03T00:00:00.000Z",
      "orders": 9,
      "platform": "Ebay",
      "revenue": 2829,
    },
    {
      "averageOrderRevenue": 303.25,
      "date": "2023-03-04T00:00:00.000Z",
      "orders": 8,
      "platform": "Ebay",
      "revenue": 2426,
    },
    {
      "averageOrderRevenue": 584.8,
      "date": "2023-03-05T00:00:00.000Z",
      "orders": 5,
      "platform": "Allegro",
      "revenue": 2924,
    },
    {
      "averageOrderRevenue": 344.4,
      "date": "2023-03-05T00:00:00.000Z",
      "orders": 5,
      "platform": "Amazon",
      "revenue": 1722,
    },
    {
      "averageOrderRevenue": 78.8,
      "date": "2023-03-05T00:00:00.000Z",
      "orders": 5,
      "platform": "Ebay",
      "revenue": 394,
    },
    {
      "averageOrderRevenue": 220,
      "date": "2023-03-09T00:00:00.000Z",
      "orders": 8,
      "platform": "Amazon",
      "revenue": 1760,
    },
    {
      "averageOrderRevenue": 212.44444,
      "date": "2023-03-09T00:00:00.000Z",
      "orders": 9,
      "platform": "Ebay",
      "revenue": 1912,
    },
  ],
}
`;

exports[`/api/platform-stats /total responds with total stats for the timerange 1`] = `
{
  "totalStats": [
    {
      "averageOrderRevenue": 309.0319892,
      "orders": 150,
      "platform": "Allegro",
      "revenue": 43213,
    },
    {
      "averageOrderRevenue": 258.840538125,
      "orders": 114,
      "platform": "Amazon",
      "revenue": 28096,
    },
    {
      "averageOrderRevenue": 260.158412173913,
      "orders": 198,
      "platform": "Ebay",
      "revenue": 54118,
    },
    {
      "averageOrderRevenue": 242.746395625,
      "orders": 131,
      "platform": "Farfetch",
      "revenue": 34267,
    },
  ],
}
`;
